iree/turbine/kernel/wave/decompose_scan_ops.py

```diff
@@ -0,0 +1,232 @@
+import math
+from operator import ge
+from typing import Any, Callable, Optional
+
+import sympy
+import torch.fx as fx
+
+from iree.turbine.kernel._support.indexing import IndexSymbol
+from iree.turbine.kernel.wave.utils.symbol_utils import subs_idxc
+
+from .._support.dtype import i1
+from .._support.tracing import CapturedTrace
+from ..ops.wave_ops import (
+    Add,
+    Cumsum,
+    CustomOp,
+    Extract,
+    NewRegister,
+    Reshape,
+    ScanOp,
+    SelectOp,
+    ShuffleOp,
+    get_custom,
+)
+from ..wave.constraints import (
+    HardwareConstraint,
+)
+from .utils.graph_utils import DCE
+
+TKW_COMBINER = {"cumsum": Add}
+IDENTITY = {"add": 0.0}
+
+
+def get_graph_node(custom: CustomOp, graph: fx.Graph) -> fx.Node:
+    custom.add_to_graph(graph)
+    return custom.fx_node
+
+
+def get_register_as_graph_node(
+    node: fx.Node,
+    value: float | sympy.Basic,
+    graph: fx.Graph,
+    shape,
+    dtype: Optional[Any] = None,
+) -> fx.Node:
+    dtype = dtype if dtype else get_custom(node).type.dtype
+    return get_graph_node(NewRegister(shape, dtype, value), graph)
+
+
+def emit_local_inclusive_scan(
+    binary_fn: Callable,
+    scan_src: fx.Node,
+    graph: fx.Graph,
+    elements_per_thread: int,
+    hardware_constraint: HardwareConstraint,
+) -> list[fx.Node]:
+    """
+    Perform local inclusive scan for `n` elements per thread.
+    """
+    elems = []
+    for i in range(elements_per_thread):
+        global_index = [i]
+        scalar = get_graph_node(Extract(scan_src, global_index), graph)
+        elems.append(scalar)
+
+    for i in range(1, elements_per_thread):
+        elems[i] = get_graph_node(binary_fn(elems[i], elems[i - 1]), graph)
+
+    return elems
+
+
+def emit_global_scan(
+    binary_fn: Callable,  # Supports only Add for now.
+    src: fx.Node,
+    local_scan: list[fx.Node],
+    graph: fx.Graph,
+    subgroup_size: int,
+    hardware_constraint: HardwareConstraint,
+    local_scan_size: int,
+    scan_dim: IndexSymbol,
+) -> fx.Node:
+    """
+    Emit an intra-warp inclusive scan using butterfly pattern scan and masking.
+    """
+    offset = local_scan[-1]
+    lane_id = (
+        hardware_constraint.linearized_thread_id % hardware_constraint.threads_per_wave
+    )
+
+    target_shape = list(src.type.symbolic_shape)
+    target_shape.pop(target_shape.index(scan_dim))
+
+    thread_incl = offset
+
+    num_steps = int(math.log2(float(subgroup_size)))
```

Comment:
currently, we are looping in a pow of 2 in range of subgroup size which is 64. 
To support, unaligned shapes, this needs to revisited.